File: dart/packages/aoc2021/lib/src/day01.dart
// https://adventofcode.com/2021/day/1

import 'package:aoc/aoc.dart';
import 'package:aoc/util/range.dart';
import 'package:aoc/util/string.dart';

main() => Day01().solve();

class Day01 extends AdventDay {
  Day01() : super(2021, 1);

  @override
  dynamic part1(String input) {
    return increases(parseMeasurements(input));
  }

  @override
  dynamic part2(String input) {
    final measurements = parseMeasurements(input);
    return increases(range(measurements.length - 2).map(
        (i) => measurements[i] + measurements[i + 1] + measurements[i + 2]));
  }

  List<int> parseMeasurements(String input) =>
      input.lines.map((s) => int.parse(s)).toList();

  int increases(Iterable<int> measurements) {
    int previous = measurements.first;
    int increases = 0;
    for (final int measurement in measurements.skip(1)) {
      if (measurement > previous) {
        increases++;
      }
      previous = measurement;
    }
    return increases;
  }
}

File: clojure/src/aoc2018/day01.clj
;; https://adventofcode.com/2018/day/1
(ns aoc2018.day01
  (:require
   [aoc.day :as d]
   [aoc.util.collection :as c]
   [aoc.util.string :as s]))

(defn input [] (d/day-input 2018 1))

(defn part1 [input] (reduce + (s/ints input)))

(defn part2 [input]
  (->> (s/ints input)
       cycle
       (reductions + 0)
       c/first-duplicate))
